kotlinx-coroutines-core/wasmWasi/src/internal/CoroutineRunner.kt

```diff
@@ -0,0 +1,13 @@
+package kotlinx.coroutines.internal
+
+import kotlinx.coroutines.*
+import kotlin.coroutines.*
+
+@InternalCoroutinesApi
+public fun runTestCoroutine(context: CoroutineContext, block: suspend CoroutineScope.() -> Unit) {
+    val newContext = GlobalScope.newCoroutineContext(context)
+    val coroutine = object: AbstractCoroutine<Unit>(newContext, initParentJob = true, active = true) {}
+    coroutine.start(CoroutineStart.DEFAULT, coroutine, block)
+    runEventLoop()
```

Comment:
Let's take a look at this test (which shouldn't be added to the codebase, except maybe as `@Ignored`):
```kotlin
    @Test
    fun testHappilyHanging() = runTest(timeout = 1.seconds) {
        // launch { // uncomment for the edge case where the job completes but its children don't
        withContext(NonCancellable) {
            println("Start hanging forever")
            Channel<Unit>().receive()
        }
        // }
    }
```

On the JVM, this test hangs, as expected: there is nothing to obtain from an empty channel, and we ignore the timeout-induced cancellation requests. On Wasm/JS, it fails with "Unknown" after a couple of seconds (ok, weird, but at least the test doesn't pass). On JS, it fails due to the Mocha 2-second timeout.

On WASI, it just exits and reports success. When the event loop has nothing more to do, it just considers the coroutine as finished.

What we can do in this case is platform-specific; for this case, I propose us to do something like
```kotlin
    check(coroutine.isCompleted) { "Coroutine $coroutine did not complete, but the system reached quiescence") }
```
directly below this line.